File: library/src/main/java/me/shkschneider/skeleton/ui/MyRecyclerViewSeparator.java
package me.shkschneider.skeleton.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

// <http://stackoverflow.com/q/24618829/>
public class MyRecyclerViewSeparator extends RecyclerView.ItemDecoration {

    public static final int HORIZONTAL = LinearLayoutManager.HORIZONTAL;
    public static final int VERTICAL = LinearLayoutManager.VERTICAL;

    private static final int[] ATTRS = new int[] {
            android.R.attr.listDivider
    };

    private Drawable mDivider;
    private int mOrientation;

    public MyRecyclerViewSeparator(final Context context, final int orientation) {
        final TypedArray typedArray = context.obtainStyledAttributes(ATTRS);
        mDivider = typedArray.getDrawable(0);
        typedArray.recycle();
        setOrientation(orientation);
    }

    public void setOrientation(final int orientation) {
        if (orientation != HORIZONTAL && orientation != VERTICAL) {
            throw new IllegalArgumentException("invalid orientation");
        }
        mOrientation = orientation;
    }

    @SuppressWarnings("deprecation")
    @Override
    public void onDraw(final Canvas canvas, final RecyclerView parent) {
        if (mOrientation == VERTICAL) {
            drawVertical(canvas, parent);
        }
        else {
            drawHorizontal(canvas, parent);
        }
    }

    public void drawVertical(final Canvas canvas, final RecyclerView parent) {
        final int left = parent.getPaddingLeft();
        final int right = parent.getWidth() - parent.getPaddingRight();
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
            final int top = child.getBottom() + params.bottomMargin;
            final int bottom = top + mDivider.getIntrinsicHeight();
            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(canvas);
        }
    }

    public void drawHorizontal(final Canvas canvas, final RecyclerView parent) {
        final int top = parent.getPaddingTop();
        final int bottom = parent.getHeight() - parent.getPaddingBottom();
        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
            final int left = child.getRight() + params.rightMargin;
            final int right = left + mDivider.getIntrinsicHeight();
            mDivider.setBounds(left, top, right, bottom);
            mDivider.draw(canvas);
        }
    }

    @Override
    public void getItemOffsets(final Rect outRect, final View view, final RecyclerView parent, final RecyclerView.State state) {
        if (parent.getChildAdapterPosition(view) == parent.getAdapter().getItemCount() - 1) {
            return;
        }
        if (mOrientation == VERTICAL) {
            outRect.set(0, 0, 0, mDivider.getIntrinsicHeight());
        }
        else {
            outRect.set(0, 0, mDivider.getIntrinsicWidth(), 0);
        }
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/ui/MyRecyclerViewSeparator.kt
package me.shkschneider.skeleton.ui

import android.graphics.Rect
import android.support.v7.widget.RecyclerView
import android.view.View

// <https://stackoverflow.com/a/29168276>
class MyRecyclerViewSeparator : RecyclerView.ItemDecoration {

    private val size: Int
    private val grid: Int
    private var spacing = false

    // Resources.getDimensionPixelSize()
    constructor(px: Int, grid: Int) {
        size = px
        this.grid = grid
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        val frameWidth = ((parent.width - size.toFloat() * (grid - 1)) / grid).toInt()
        val padding = parent.width / grid - frameWidth
        val itemPosition = (view.layoutParams as RecyclerView.LayoutParams).viewAdapterPosition
        if (itemPosition < grid) {
            outRect.top = 0
        } else {
            outRect.top = size
        }
        if (itemPosition % grid == 0) {
            outRect.left = 0
            outRect.right = padding
            spacing = true
        } else if ((itemPosition + 1) % grid == 0) {
            spacing = false
            outRect.right = 0
            outRect.left = padding
        } else if (spacing) {
            spacing = false
            outRect.left = size - padding
            if ((itemPosition + 2) % grid == 0) {
                outRect.right = size - padding
            } else {
                outRect.right = size / 2
            }
        } else if ((itemPosition + 2) % grid == 0) {
            spacing = false
            outRect.left = size / 2
            outRect.right = size - padding
        } else {
            spacing = false
            outRect.left = size / 2
            outRect.right = size / 2
        }
        outRect.bottom = 0
    }

}
